### PR TITLE
Get rid of DISPOSE_SWAL_TIMEOUT

### LINKED ISSUE #1654: After v8.13.2, Angular not becoming "stable" after closing Swal.
## Current behavior

After v8.13.2, in an Angular application (tested in Angular 7.2 nd 8.0), fire a Swal and close it. In browser console run...

window.getAllAngularTestabilities()[0].isStable();

This will report false until the DISPOSE_SWAL_TIMEOUT is reached.

## Expected behavior

Once the Swal is closed, isStable() should return true. This is currently breaking our selenium tests as we don't proceed with a lot of actions until Angular reports back "stable."

### PR DESCRIPTION
**Pull request by limonte**
Fixes #1654

I moved `disposeSwal()` to `removePopupAndResetState()`. @gverni let me know if I'm missing some use-cases.

No tests are failing after removing `DISPOSE_SWAL_TIMEOUT` so either we don't have that functionality covered by tests or everything is well without `DISPOSE_SWAL_TIMEOUT`.

@gverni please take a look and let me know what do you think

### CODE REVIEW

**Comment by limonte** [comment]
If there's some use-cases I'm missing here, it would be better to use the programming language and add the test case which is failing to this PR instead of explaining it with humans language :)

**Comment by gverni** [comment]
@limonte the missing test that _fail_ is the following:

```javascript
  Swal.fire({
    input: 'text',
    inputClass: 'outer-swal',
  }).then((result) => {
    setTimeout(() => {
      console.log(Swal.getInput().classList)
    },160)
  })
```

See also [this comment](https://github.com/sweetalert2/sweetalert2/issues/1640#issuecomment-507383036)

The reason why I did not add this test, is because for me the above code just show that the error caused by it is not _deterministic_. Basically if the code in the `then` takes more than the animation (the duration of which is customizable) then we hit that issue. 

Having said that... I agree with 200% that I should have add the above test at least for the _default_ animation.

**Comment by limonte** [comment]
```js
  Swal.fire({
    input: 'text',
    inputClass: 'outer-swal',
  }).then((result) => {
    setTimeout(() => {
      console.log(Swal.getInput().classList)
    },160)
  })
```

What's the expected output in your opinion here and why?

**Comment by gverni** [comment]
> What's the expected output in your opinion here and why?

As you said, a piece of code is worth a thousands words :smile:, so let me put it in a test format: 

```javascript

  Swal.fire({
    input: 'text',
    inputClass: 'outer-swal',
  }).then((result) => {
    var inputClassList = Swal.getInput().classList)
    setTimeout(() => {
      asert.equal(inputClassList,  Swal.getInput().classList)
    },160)
  })

```

**Comment by gverni** [comment]
To clarify my test, a statement should not return a different value (or triggering an error) just depending on when it gets executed. In my test code I put a timeout, but that could have been a `fetch()` in a real use case. I know that my solution with the timeout didn't fully solve the problem, but at least is giving enough time (maybe too much) for any code to be executed.

**Comment by limonte** [comment]
> a statement should not return a different value (or triggering an error) just depending on when it gets executed. 

This statement can't be valid. Currently (with `DISPOSE_SWAL_TIMEOUT` the same value will be for 36 seconds. We can change the magical number to something else but that would still be a bad idea :)

> In my test code I put a timeout, but that could have been a fetch() in a real use case. I know that my solution with the timeout didn't fully solve the problem, but at least is giving enough time (maybe too much) for any code to be executed.

I still can't understand your reasoning, why a developer can't do something like:

```js
  Swal.fire({
    ...
  }).then((result) => {
    var inputValue = Swal.getInput().value
    fetch('http://example.com/movies.json')
    .then(function(response) {
      console.log(inputValue)
    })
  })
```

?

**Comment by gverni** [comment]
> I still can't understand your reasoning, why a developer can't do something like:

They certainly can. But what if they develop and test using: 

```javascript
  Swal.fire({
    ...
  }).then((result) => {
    fetch('http://example.com/movies.json')
    .then(function(response) {
      console.log(Swal.getInput().value)
    })
  })

```
In test environment everything works fine (e.g. the fetch is done on a local test server). But then in production the `fetch` take more than 160ms. They will be scratching their head thinking why the code works in a different way. What is important for me is to have something deterministic that can be documented. If we use the timeout, we can add that to the documentation so everybody know that after the timeout expires, then the swal objects are disposed.

**Comment by limonte** [comment]
> In test environment everything works fine (e.g. the fetch is done on a local test server). But then in production the fetch take more than 160ms. They will be scratching their head thinking why the code works in a different way.

Scratching heads is good for learning how to write a robust code which works the same way in all environments :) The code example is bad, let's not support bad-written use cases just because someone can come up with them.

> What is important for me is to have something deterministic that can be documented.

That's exactly my point. Whaever magic number for the timeout we will come up with, there will be a newby with fetch taking `timeout+1s` :)

**Comment by gverni** [comment]
@limonte your current code is not behaving correctly with `animation: false`. 

Whereas: 
```javascript
  Swal.fire({
    text: 'hello',
    input: 'text',
    animation: false
  }).then(() => {
    console.log(Swal.getInput())
  })
```
Produces: `null`
This is a breaking change, and that means developer are not able to access to Swal objects if there is no animation. 

Also, this is different from what when the animation is present: 

```javascript
  Swal.fire({
    text: 'hello',
    input: 'text',
  }).then(() => {
    console.log(Swal.getInput())
  })
```
Produces: `<input class="swal2-input" ...`

Are you sure about that?

**Comment by gverni** [comment]
> This is a breaking change and also means developers are not able to access to Swal objects if there is no animation.

The reason why `animation: false` was working before version 8.13.2 is explained in [this comment](https://github.com/sweetalert2/sweetalert2/issues/1640#issuecomment-507347802).

**Comment by gverni** [comment]
@limonte let me try to summarise the requirements that brought us to this PR: 

**Requirements**
1. In the fix for #1569 disposal of the Swal internal objects (`privateProps` and `privateMethods`) was added to allow garbage collector to free their memory
2. Since the developers need to access the Swal private properties (e.g. `Swal.getInput()`) in the `then`, those private properties cannot be disposed before the code in the `then` is executed

**Solution before v1.8.2**
* To satisfy requirement 2., a delay before disposing Swal internal objects needed to be introduced
* The Swal animation was used as this delay (~150ms): disposal was done in the handling of animation end event (i.e. `animationEndEvent`)
* In case Swal had no animation, no disposal was done (effectively non satisfying requirement 1.)
* If the animation was interrupted, the _animation end_ event was not triggered, with a risk of not satisfying requirement 1

**Solution introduced in v1.8.2**
* A fixed timeout is introduced, controlled by constant `DISPOSE_SWAL_TIMEOUT` 
* The disposal of the Swal internal objects is executed regardless of the Swal animation
* Side effect: this creates the issue #1654 

Honestly I don't know how to satisfy both requirement 1 and 2 without introducing a timeout (either in the form of `DISPOSE_SWAL_TIMEOUT` or the animation end event). I tried to see if there is a way to execute some code after the code in the `then` finishes to execute (e.g. `finally`?) but I couldn't develop anything working so far.

**Comment by limonte** [comment]
> @limonte your current code is not behaving correctly with `animation: false`.
> 
> ```js
>   Swal.fire({
>     text: 'hello',
>     input: 'text',
>     animation: false
>   }).then(() => {
>     console.log(Swal.getInput())
>   })
> ```
> 
> Produces: `null`
> This is a breaking change and also means developers are not able to access to Swal objects if there is no animation.
> 
> Also, this is different from the behavior when the animation is present:
> 
> ```js
>   Swal.fire({
>     text: 'hello',
>     input: 'text',
>   }).then(() => {
>     console.log(Swal.getInput())
>   })
> ```
> 
> Produces: `<input class="swal2-input" ...`
> 
> Are you sure about that?

That's a good catch, thank you @gverni! We are missing some important test cases. Done in https://github.com/sweetalert2/sweetalert2/pull/1655/commits/b3b651f025974b7baabc0c2e0ed44405e3457482

Any objections?

File: src/instanceMethods/close.js [hunk 1]
```diff
@@ -117,12 +118,13 @@ const unsetWeakMaps = (obj) => {
...
-  if (onAfterClose !== null && typeof onAfterClose === 'function') {
-    setTimeout(() => {
+const triggerOnAfterCloseAndDispose = (instance, onAfterClose) => {
+  setTimeout(() => {
+    if (onAfterClose !== null && typeof onAfterClose === 'function') {
       onAfterClose()
-    })
-  }
+    }
+    disposeSwal(instance)
```
Comment (gverni) [inline src/instanceMethods/close.js]: The `disposeSwal()` executes after the `onAfterClose()`. If the `onAfterClose()` contains a new Swal, then the inner properties of the new Swal get disposed. I added a new test for that in https://github.com/sweetalert2/sweetalert2/pull/1655/commits/9bed120a28d80e346bda7c638f6f578bbec05dea

Comment (limonte) [reply src/instanceMethods/close.js]: Good catch and thanks a lot for providing a test case! A failing test case worth thousands of words in GitHub comments :)

The provided test case is not failing with the check I made in the latest commit: https://github.com/sweetalert2/sweetalert2/pull/1655/commits/4510d138ab31beca73292d405b8b5abade33d365

**Comment by gverni** [comment]
@limonte it looks good to me now, but can you give me a couple of days more? I’m out this weekend and I’m not able to try some more corner cases. Is that ok if I finish the review on Monday? If you are in a rush you can go ahead and merge for now, or maybe some other collaborator can chime in :wink:

**Comment by limonte** [comment]
No hurry @gverni, this PR can wait till Monday, enjoy the weekend!

**Comment by gverni** [comment]
@limonte I did some more tests, and everything looks good. I added two commits to this PR: 

* 504867b: I've seen you used `dom.getPopup()` to check if a Swal is open. I feel this is a much better way than using a global variable. Hence I'm removing `globalState.swalClosing` in favor of that
* 8da72d2: added one test to make sure that when a Swal is fired in the `onClose` function, then this Swal is accessible in the `then`

This latter has more to do with the `onClose` / `onAfterClose` sequence, and I think it's useful in the context of this PR to avoid any breaking changes. 

I was thinking to create more tests based on the use cases I tested for this PR. I may do that once this PR is merged if that's ok for you.

**Comment by limonte** [comment]
Thank you @gverni!!

**Pull request merged**